pkg/services/notary/notary.go

```diff
@@ -448,33 +477,37 @@ func (n *Notary) newTxCallbackLoop() {
 		case tx := <-n.newTxs:
 			isMain := tx.tx.Hash() == tx.mainHash
 
-			n.reqMtx.Lock()
+			n.reqMtx.RLock()
 			r, ok := n.requests[tx.mainHash]
-			if !ok || isMain && (r.isSent || r.minNotValidBefore <= n.Config.Chain.BlockHeight()) {
-				n.reqMtx.Unlock()
+			n.reqMtx.RUnlock()
+			if !ok {
+				continue
+			}
+			r.lock.Lock()
```

Comment:
Suggest moving the code below to some helper to have a single lock/unlock. It should work now, but it's easy to miss some unlock if changing anything here.